{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2dd67df3-178c-4355-9ce2-bec6d808b465",
   "metadata": {},
   "source": [
    "# Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3c8cfe2-2f43-443d-bd23-5d1e97c5f5fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(rgl.useNULL=TRUE)\n",
    "library(rgl)\n",
    "library(usethis)\n",
    "library(lattice)\n",
    "library(adespatial)\n",
    "library(vegan)\n",
    "library(devtools)\n",
    "library(phyloseq)\n",
    "library(microbiomeSeq)\n",
    "library(DECIPHER)\n",
    "library(ape)\n",
    "library(GenomicRanges)\n",
    "library(SummarizedExperiment)\n",
    "library(MatrixGenerics)\n",
    "library(matrixStats)\n",
    "library(DESeq2)\n",
    "library(ggplot2)\n",
    "library(plotly)\n",
    "library(philr)\n",
    "library(tidyverse)\n",
    "library(adespatial)\n",
    "library(qiime2R)\n",
    "library(MicrobeR)\n",
    "library(microbiome)\n",
    "library(ShortRead)\n",
    "library(microbiomeSeq)\n",
    "library(pander)\n",
    "library(ranacapa)\n",
    "library(grid)\n",
    "library(gridExtra)\n",
    "library(knitr)\n",
    "library(png)\n",
    "library(heplots)\n",
    "library(ggdendro)\n",
    "library(ggpubr)\n",
    "library(RColorBrewer)\n",
    "library(microbiomeutilities)\n",
    "library(viridis)\n",
    "library(magrittr)\n",
    "library(dplyr)\n",
    "library(ggstatsplot)\n",
    "library(dada2)\n",
    "library(phangorn)\n",
    "library(randomcoloR)\n",
    "library(Polychrome)\n",
    "library(microbiome)\n",
    "library(remotes)\n",
    "library(upstartr)\n",
    "library(hrbrthemes)\n",
    "library(psadd)\n",
    "library(Seurat)\n",
    "library(microbiomer)\n",
    "library(VennDiagram)\n",
    "library(ggvenn)\n",
    "library(ggprism)\n",
    "library(multiplyr)\n",
    "library(rstatix)\n",
    "library(janitor)\n",
    "library(microbiome)\n",
    "library(microViz)\n",
    "library(Rtsne)\n",
    "library(ggh4x)\n",
    "library(RColorBrewer)\n",
    "library(reshape2)\n",
    "library(zipangu)\n",
    "library(grDevices)\n",
    "library(ragg)\n",
    "library(ALDEx2)\n",
    "library(seqinr)\n",
    "library(seqateurs)\n",
    "library(metagMisc)\n",
    "library(data.table)\n",
    "library(ggpicrust2)\n",
    "library(Maaslin2)\n",
    "library(KEGGREST)\n",
    "library(PERFect)\n",
    "library(packrat)\n",
    "library(NCmisc)\n",
    "library(plyr)\n",
    "library(ggVennDiagram)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47785721-2aa3-4065-bb41-3a4f68c58cfa",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Definitions, R Image & Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccdf3c0-3ac3-4ef6-b452-bbe945a3d796",
   "metadata": {},
   "source": [
    "### Definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd731ad8-42c4-44b6-b3fa-0c7858afcfce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path <- \"/Users/dzakaria/research/GiantTrotoisesMicrobiome/research/data/16s/tartarughe_2022/data_processing/\" \n",
    "R_image <- paste(path,\"20231021-AldabraGiantTortoises-dada2.RData\", sep=\"\")\n",
    "EnvironmentGroup5_sorted <- c(\"Curieuse\\n(Nursery)\", \"Curieuse\\n(Perley)\", \"Curieuse\", \"Botanical\\nGarden\", \"Parco\\nNatura\")\n",
    "AgeGroup13_sorted <- c(\"1\", \"2\", \"4\", \"5\", \"15\", \"30\", \"40\", \"50\", \"60\", \"70\", \"80\", \"90\", \"100\")\n",
    "AgeGroup_sorted <- c(\"Young\\n(0-30)\", \"Mid Age\\n(30-70)\", \"Old\\n(70-100)\", \"Centenarian\\n(>100)\")\n",
    "AgeGroup_vennsorted <- c(\"Mid Age\\n(30-70)\", \"Old\\n(70-100)\", \"Centenarian\\n(>100)\", \"Young\\n(0-30)\")\n",
    "AgeGroup_vennsorted_wild <- c(\"Centenarian\\n(>100)\",\"Mid Age\\n(30-70)\",\"Old\\n(70-100)\")\n",
    "AgeGroup_vennsorted_controlled <- c(\"Old\\n(70-100)\",\"Centenarian\\n(>100)\",\"Young\\n(0-30)\", \"Mid Age\\n(30-70)\")\n",
    "EnvironmentGroup5_vennsorted <- c(\"Parco\\nNatura\", \"Curieuse\",\"Curieuse\\n(Perley)\",\"Botanical\\nGarden\", \"Curieuse\\n(Nursery)\")\n",
    "\n",
    "## Preparing color palettes\n",
    "mycols_environment <- c(\"olivedrab\", \"olivedrab3\", \"orange4\")\n",
    "mycols_environment2 <- c(\"orange4\", \"orange3\", \"olivedrab\")\n",
    "mycols_environment4 <- c(\"orange3\", \"orange4\", \"olivedrab\", \"olivedrab2\")\n",
    "mycols_environment5 <- c(\"orange2\",\"orange3\", \"orange4\", \"olivedrab\", \"olivedrab2\")\n",
    "mycols_agegroup13 <- colorRampPalette(c(\"wheat3\",\"salmon2\"))(13)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "921167d7-b02b-41ae-af98-c9fe403060ba",
   "metadata": {},
   "source": [
    "### Load R Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2791aa61-d531-448e-8913-8f1a7b715d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The directory containing fastq files after unzipping\n",
    "setwd(path)\n",
    "load(file = R_image)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa94004a-d9dd-4557-9b68-31d3310f171a",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "628e18c0-c4e8-47d1-8d67-c3051cd6e1a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha_diversity_calculation <- \n",
    "  function(ps){\n",
    "  ps_df <- phyloseq::psmelt(ps)\n",
    "  ps_rich <- estimate_richness(ps)\n",
    "  row.names(ps_rich) <- gsub(\"X\", \"\", row.names(ps_rich))\n",
    "  ps_rich$Sample <- row.names(ps_rich)\n",
    "  ps_df_rich <- merge(ps_rich, ps_df, by = \"Sample\")\n",
    "  ps_df_rich$AgeGroup13 <- as.factor(ps_df_rich$AgeGroup13)\n",
    "  return(ps_df_rich)\n",
    "}\n",
    "\n",
    "alpha_diversity_plot <- \n",
    "  function(ps_df_rich, xname, yname, ref_group_wilcox, jitter_colors_variable_name, jitter_colors, x_axis_lab, y_axis_lab, plot_title, legend_title){\n",
    "  xdata <- ps_df_rich[,xname]\n",
    "  ydata <- ps_df_rich[,yname]\n",
    "  jitter_colors_variabledata <- ps_df_rich[,jitter_colors_variable_name]\n",
    "  ggplot(ps_df_rich, aes(x = xdata, y = ydata)) +\n",
    "    geom_boxplot()+\n",
    "    geom_jitter(position = position_dodge(0.8), cex = 3, aes(color=jitter_colors_variabledata)) +\n",
    "    labs(x = x_axis_lab, y= y_axis_lab, title = stringr::str_wrap(plot_title, width = 50)) +\n",
    "    theme_prism() +\n",
    "    scale_color_manual(legend_title, values = jitter_colors) +\n",
    "    theme(\n",
    "      plot.margin = margin(1, 1, 1, 1, \"cm\"),\n",
    "      legend.position = \"right\",\n",
    "      axis.text.x = element_text(color = \"black\", size = 25),\n",
    "      axis.text.y = element_text(color = \"black\", size = 25),\n",
    "      axis.title = element_text(size = 30),\n",
    "      legend.title = element_text(size = 30),\n",
    "      legend.text = element_text(size = 25),\n",
    "      strip.text.x = element_text(size = 28),\n",
    "      plot.title = element_text(size = 40)) +\n",
    "    stat_compare_means(\n",
    "      aes(label = after_stat(p.signif)),\n",
    "      method = \"wilcox.test\",\n",
    "      ref.group = ref_group_wilcox)\n",
    "  }\n",
    "\n",
    "ps_rel_abund_taxacore_createdf <- function(ps_rel_abun, taxa_name){\n",
    "  taxa_name_capitalized <- str_to_title(taxa_name)\n",
    "  if(taxa_name_capitalized == \"Phylum\" | taxa_name_capitalized == \"Species\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"prevalence\" =as.integer(0))\n",
    "  } else {\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"prevalence\" =as.integer(0.5))\n",
    "  }\n",
    "  ps_rel_abund_taxacore <- aggregate_rare(ps_rel_abund, taxa_variables[\"taxa_name_capitalized\"], detection = 0, prevalence = taxa_variables[\"prevalence\"])\n",
    "  ps_rel_abund_taxacore_df <- phyloseq::psmelt(ps_rel_abund_taxacore)\n",
    "  ps_rel_abund_taxacore_df$AgeGroup13 = factor(ps_rel_abund_taxacore_df$AgeGroup13, levels = AgeGroup13_sorted)\n",
    "  ps_rel_abund_taxacore_df$EnvironmentGroup5 = factor(ps_rel_abund_taxacore_df$EnvironmentGroup5, levels = EnvironmentGroup5_sorted)\n",
    "  ps_rel_abund_taxacore_df$Sample <- factor(ps_rel_abund_taxacore_df$Sample, levels = unique(ps_rel_abund_taxacore_df$Sample[order(ps_rel_abund_taxacore_df$AgeGroup13)]))\n",
    "  return(ps_rel_abund_taxacore_df)\n",
    "}\n",
    "\n",
    "ps_rel_abund_taxacore_plot <- function(ps_rel_abund_taxacore_df,taxa_name){\n",
    "taxa_name_capitalized <- str_to_title(taxa_name)\n",
    "  if(taxa_name_capitalized == \"Phylum\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_phylum\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Phyla in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  } else if(taxa_name_capitalized == \"Family\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_family\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Families in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  }else if(taxa_name_capitalized == \"Genus\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_genus\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Genera in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  }else if(taxa_name_capitalized == \"Species\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_species\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Species in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  }\n",
    "  if(!(exists(as.character(taxa_variables[\"col_set\"])))){\n",
    "    taxa <- unique(ps_rel_abund_taxacore_df[, as.character(taxa_variables[\"taxa_name_capitalized\"])])\n",
    "    col_set <- data.frame(taxa, distinctColorPalette(length(taxa)))\n",
    "    col_set <- rbind(col_set, c(\"Remainder\", \"grey\"))\n",
    "    rownames(col_set) <- NULL\n",
    "    colnames(col_set) <- c(as.character(taxa_variables[\"taxa_name_capitalized\"]), \"Color\")\n",
    "    class(col_set)\n",
    "    col_set <- setNames(as.character(col_set[,2]), col_set[,1])\n",
    "    assign(as.character(taxa_variables[\"col_set\"]), col_set, envir = .GlobalEnv)\n",
    "  print(\"you created col set\")\n",
    "  }\n",
    "  print(\"you already had col set\")\n",
    "  ggplot(ps_rel_abund_taxacore_df,\n",
    "       aes(x = Sample, \n",
    "           Abundance, \n",
    "           fill = fct_reorder(eval(parse(text = taxa_variables[\"taxa_name_capitalized\"])), Abundance, .desc = TRUE))) +\n",
    "      labs(x = \"Age Group\", \n",
    "           y = \"Relative Abundance (%)\", \n",
    "           title = stringr::str_wrap(taxa_variables[\"plot_title\"], width = 50))+\n",
    "      facet_grid(cols = vars(EnvironmentGroup5), \n",
    "                 scales = \"free\", \n",
    "                 space = \"free\") +\n",
    "      theme_bw(base_family = \"HiraKakuProN-W3\") +\n",
    "      theme_prism() +\n",
    "      geom_bar(stat = \"identity\", \n",
    "               position = \"fill\")+\n",
    "      scale_fill_manual(as.character(taxa_variables[\"taxa_name_capitalized\"]),\n",
    "                        values = eval(parse(text = taxa_variables[\"col_set\"]))) +\n",
    "      scale_y_percent(scale = 100) +\n",
    "      scale_x_discrete(breaks = ps_rel_abund_taxacore_df$Sample, labels = ps_rel_abund_taxacore_df$AgeGroup13) +\n",
    "      theme(\n",
    "        plot.margin = margin(1, 1, 1, 1, \"cm\"),\n",
    "        axis.text.x = element_text(angle = 90, color = \"black\", size = 18),\n",
    "        axis.text.y = element_text(color = \"black\", size = 20),\n",
    "        axis.title = element_text(size = 25),\n",
    "        legend.position = \"bottom\",\n",
    "        legend.title = element_text(size = 20, hjust=0),\n",
    "        legend.text = element_text(size = 12),\n",
    "        strip.text.x = element_text(size = 12),\n",
    "        plot.title = element_text(size = 40, hjust=0.5),\n",
    "        panel.spacing = unit(1, \"lines\")) +\n",
    "      guides(fill = guide_legend(nrow = 9))  \n",
    "  \n",
    "}\n",
    "\n",
    "ps_rel_abund_taxacore_0.01remainder_createdf_plot <- function(ps_rel_abund_taxacore_df, taxa_name){\n",
    "  taxa_name_capitalized <- str_to_title(taxa_name)\n",
    "    if(taxa_name_capitalized == \"Phylum\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_phylum\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Phyla (with >1% Abundance) in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  } else if(taxa_name_capitalized == \"Family\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_family\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Families (with >1% Abundance) in the Gut Microbiome of Aldabra Giant Tortoises \")\n",
    "  }else if(taxa_name_capitalized == \"Genus\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_genus\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Genera (with >1% Abundance) in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  }else if(taxa_name_capitalized == \"Species\"){\n",
    "    taxa_variables <- c(\n",
    "      \"taxa_name_capitalized\" = taxa_name_capitalized,\n",
    "      \"col_set\" =\"col_species\",\n",
    "      \"plot_title\" = \"Relative Abundance of Bacterial Species (with >1% Abundance) in the Gut Microbiome of Aldabra Giant Tortoises\")\n",
    "  }\n",
    "  ps_rel_abund_taxacore_0.1remainder_df <- ps_rel_abund_taxacore_df\n",
    "  ps_rel_abund_taxacore_0.1remainder_df[,as.character(taxa_variables[\"taxa_name_capitalized\"])] <- as.character(ps_rel_abund_taxacore_0.1remainder_df[,as.character(taxa_variables[\"taxa_name_capitalized\"])])\n",
    "  medians <- ddply(ps_rel_abund_taxacore_0.1remainder_df, as.quoted(as.character(taxa_variables[\"taxa_name_capitalized\"])), function(x) c(median=median(x$Abundance)))\n",
    "  remainder <- c(medians[medians$median <= 0.1,as.character(taxa_variables[\"taxa_name_capitalized\"])], \"Unknown\", \"Other\")\n",
    "  ps_rel_abund_taxacore_0.1remainder_df[ps_rel_abund_taxacore_0.1remainder_df[,as.character(taxa_variables[\"taxa_name_capitalized\"])] %in% remainder,as.character(taxa_variables[\"taxa_name_capitalized\"])] <- 'Remainder'\n",
    "  ps_rel_abund_taxacore_0.1remainder_df$EnvironmentGroup5 = factor(ps_rel_abund_taxacore_0.1remainder_df$EnvironmentGroup5,\n",
    "                                                                     levels = EnvironmentGroup5_sorted)\n",
    "  ps_rel_abund_taxacore_0.1remainder_df$Sample <- factor(ps_rel_abund_taxacore_0.1remainder_df$Sample, levels = unique(ps_rel_abund_taxacore_0.1remainder_df$Sample[order(ps_rel_abund_taxacore_0.1remainder_df$AgeGroup13)]))\n",
    "\n",
    "  if(!(exists(as.character(taxa_variables[\"col_set\"])))){\n",
    "    taxa <- unique(ps_rel_abund_taxacore_df[, as.character(taxa_variables[\"taxa_name_capitalized\"])])\n",
    "    col_set <- data.frame(taxa, distinctColorPalette(length(taxa)))\n",
    "    col_set <- rbind(col_set, c(\"Remainder\", \"grey\"))\n",
    "    rownames(col_set) <- NULL\n",
    "    colnames(col_set) <- c(as.character(taxa_variables[\"taxa_name_capitalized\"]), \"Color\")\n",
    "    class(col_set)\n",
    "    col_set <- setNames(as.character(col_set[,2]), col_set[,1])\n",
    "    assign(as.character(taxa_variables[\"col_set\"]), col_set, envir = .GlobalEnv)\n",
    "  print(\"you created col set\")\n",
    "  }\n",
    "  print(\"you already had col set\")\n",
    "  \n",
    "  ggplot(ps_rel_abund_taxacore_0.1remainder_df,\n",
    "       aes(x = Sample, \n",
    "           Abundance, \n",
    "           fill = fct_reorder(eval(parse(text = taxa_variables[\"taxa_name_capitalized\"])), Abundance, .desc = TRUE))) +\n",
    "      labs(x = \"Age Group\", \n",
    "           y = \"Relative Abundance (%)\", \n",
    "           title = stringr::str_wrap(taxa_variables[\"plot_title\"], width = 60))+\n",
    "      facet_grid(cols = vars(EnvironmentGroup5), \n",
    "                 scales = \"free\", \n",
    "                 space = \"free\") +\n",
    "      theme_bw(base_family = \"HiraKakuProN-W3\") +\n",
    "      theme_prism() +\n",
    "      geom_bar(stat = \"identity\", \n",
    "               position = \"fill\")+\n",
    "      scale_fill_manual(as.character(taxa_variables[\"taxa_name_capitalized\"]),\n",
    "                        values = eval(parse(text = taxa_variables[\"col_set\"]))) +\n",
    "      scale_y_percent(scale = 100) +\n",
    "      scale_x_discrete(breaks = ps_rel_abund_taxacore_df$Sample, labels = ps_rel_abund_taxacore_df$AgeGroup13) +\n",
    "        theme(\n",
    "          plot.margin = margin(1, 1, 1, 1, \"cm\"),\n",
    "          axis.text.x = element_text(angle = 90, color = \"black\", size = 15),\n",
    "          axis.text.y = element_text(color = \"black\", size = 15),\n",
    "          axis.title = element_text(size = 20),\n",
    "          legend.position = \"right\",\n",
    "          legend.title = element_text(size = 15, hjust=0),\n",
    "          legend.text = element_text(size = 10),\n",
    "          strip.text.x = element_text(size = 11),\n",
    "          plot.title = element_text(size = 25, hjust=0.5),\n",
    "          panel.spacing = unit(1, \"lines\")) +\n",
    "        guides(fill = guide_legend(ncol = 1))  \n",
    "}\n",
    "\n",
    "beta_ordination_plot <- function(ps_rel_abund, method, distance){\n",
    "  sample_data(ps_rel_abund)$AgeGroup13 <- factor(sample_data(ps_rel_abund)$AgeGroup13, levels = AgeGroup13_sorted)\n",
    "  sample_data(ps_rel_abund)$EnvironmentGroup5 <- factor(sample_data(ps_rel_abund)$EnvironmentGroup5, levels = EnvironmentGroup5_sorted)\n",
    "  method_capitalized <- str_to_upper(method)\n",
    "  distance_capitalized <- str_to_title(distance)\n",
    "  if(method_capitalized == \"PCOA\"){method_capitalized = \"PCoA\"} else if(method_capitalized == \"DPCOA\"){method_capitalized = \"DPCoA\"}\n",
    "  if(distance_capitalized == \"Bray Curtis\"){\n",
    "    distance_variables <- c(\n",
    "      \"distance_capitalized\" = distance_capitalized,\n",
    "      \"distance\" =\"bray\")\n",
    "  } else if(distance_capitalized == \"Unweighted Unifrac\") {\n",
    "    distance_variables <- c(\n",
    "      \"distance_capitalized\" = distance_capitalized,\n",
    "      \"distance\" =\"unifrac\")    \n",
    "  } else if(distance_capitalized == \"Weighted Unifrac\") {\n",
    "    distance_variables <- c(\n",
    "      \"distance_capitalized\" = distance_capitalized,\n",
    "      \"distance\" =\"wunifrac\")    \n",
    "  }\n",
    "  ord_pcoa <- ordinate(ps_rel_abund, method = method_capitalized, distance = as.character(distance_variables['distance']))\n",
    "  plot_ordination(ps_rel_abund, ord_pcoa, \n",
    "                  color = \"AgeGroup13\", \n",
    "                  shape = \"EnvironmentGroup5\", \n",
    "                  title = paste(method_capitalized, \" plot of Beta diversity (\", as.character(distance_variables['distance_capitalized']),\")\\nof Aldabra Giant Tortoises\", sep = \"\")) +\n",
    "    theme_prism() +\n",
    "    geom_point(size = 2.3) +\n",
    "    scale_shape_manual(values = c(20, 16, 19, 17, 15)) +\n",
    "    labs(shape = \"Envrionment\", color = \"Age Group\") +\n",
    "    scale_color_manual(values = mycols_agegroup13) +\n",
    "    theme(\n",
    "      axis.text.x = element_text(color = \"black\", size = 25),\n",
    "      axis.text.y = element_text(color = \"black\", size = 25),\n",
    "      axis.title = element_text(size = 30),\n",
    "      legend.title = element_text(size = 25),\n",
    "      legend.text = element_text(size = 20),\n",
    "      strip.text.x = element_text(size = 28),\n",
    "      plot.title = element_text(size = 30)) +\n",
    "    stat_ellipse(linetype = 2)\n",
    "}\n",
    "\n",
    "venn_taxa_plot <- function(ps_rel_abund, groupby, groupby_order,taxa_name){\n",
    "  groups <- unique(as.character(microbiome::meta(ps_rel_abund)[,as.character(groupby)]))\n",
    "  groups <- factor(groups, levels = groupby_order)\n",
    "  list_core <- c()\n",
    "  if(taxa_name == \"ASVs\"){\n",
    "    for (i in groups){\n",
    "      ps.sub <- do.call(\"subset_samples\", list(quote(ps_rel_abund), substitute(eval(parse(text = groupby)) == i)))\n",
    "      core_m <- core_members(ps.sub,\n",
    "                          detection = 0.001,\n",
    "                          prevalence = 0.75)\n",
    "      list_core[[i]] <- core_m \n",
    "    }\n",
    "  } else {\n",
    "    for (i in groups){\n",
    "      ps.sub <- do.call(\"subset_samples\", list(quote(ps_rel_abund), substitute(eval(parse(text = groupby)) == i)))\n",
    "      core_m <- psmelt(aggregate_rare(ps.sub, level = taxa_name, detection = 0, prevalence = 0))\n",
    "      print(length(core_m[,taxa_name]))\n",
    "      list_core[[i]] <- core_m[,taxa_name]\n",
    "    }\n",
    "  }\n",
    "  summary(list_core)\n",
    "  list_core <- list_core[order(levels(groups))]\n",
    "  ggVennDiagram(list_core, \n",
    "                label_alpha = 0, \n",
    "                label = \"percent\", \n",
    "                edge_size = 0)+\n",
    "    ggtitle(paste(\"Venn Diagram for \", taxa_name, sep=\"\")) +\n",
    "    theme(\n",
    "      plot.margin = margin(1, 1, 1, 1, \"cm\"),\n",
    "      plot.title = element_text(hjust = 0.5, size = 20, face= \"bold\"),\n",
    "      legend.position = \"none\")+\n",
    "    scale_fill_gradient(low=\"gray100\",high = \"gray50\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29c4366-b89c-4092-8298-a716bdb665a9",
   "metadata": {},
   "source": [
    "# Create Phyloseq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d912d927-edff-413b-952d-ad11c7c2268e",
   "metadata": {},
   "source": [
    "### Preprocess Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "867a655e-a5ba-44c0-b8e7-0e6336ba1a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data processing referring to https://benjjneb.github.io/dada2/tutorial.html, https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html\n",
    "\n",
    "# Primers have already been trimmed using cutadapt\n",
    "# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq\n",
    "fnFs <- sort(list.files(path, pattern = \"_R1_001.fastq\", full.names = TRUE))\n",
    "fnRs <- sort(list.files(path, pattern = \"_R2_001.fastq\", full.names = TRUE))\n",
    "sample.names <- sapply(strsplit(basename(fnFs), \"_\"), `[`, 1)\n",
    "\n",
    "# Place filtered files in filtered/ subdirectory\n",
    "filtFs <- file.path(path, \"filtered\", paste0(sample.names, \"_F_filt.fastq.gz\"))\n",
    "filtRs <- file.path(path, \"filtered\", paste0(sample.names, \"_R_filt.fastq.gz\"))\n",
    "names(filtFs) <- sample.names\n",
    "names(filtRs) <- sample.names\n",
    "\n",
    "# Too few reads passed the filter when using maxEE = c(2,2) so it was relaxed on the reverse reads to maxEE = c(2,5)\n",
    "out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,\n",
    "                     maxN = 0, maxEE = c(2, 5), truncQ = 2, rm.phix = TRUE,\n",
    "                     compress = TRUE, multithread = TRUE\n",
    ")\n",
    "head(out)\n",
    "\n",
    "# Learn the sequencing error rates\n",
    "errF <- learnErrors(filtFs, multithread=TRUE)\n",
    "errR <- learnErrors(filtRs, multithread=TRUE)\n",
    "\n",
    "dadaFs <- dada(filtFs, err = errF, multithread = TRUE)\n",
    "dadaRs <- dada(filtRs, err = errR, multithread = TRUE)\n",
    "mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose = TRUE)\n",
    "\n",
    "# Constructing ASV table\n",
    "seqtab <- makeSequenceTable(mergers)\n",
    "table(nchar(getSequences(seqtab)))\n",
    "\n",
    "# The expected amplicon size is 260 ~ 460 bp\n",
    "seqtab <- seqtab[,nchar(colnames(seqtab)) %in% 240:460]\n",
    "\n",
    "# Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant “parent” sequences\n",
    "seqtab.nochim <- removeBimeraDenovo(seqtab, method = \"consensus\", multithread = TRUE, verbose = TRUE)\n",
    "sum(seqtab.nochim)/sum(seqtab)\n",
    "# Tracking reads throught the pipeline\n",
    "getN <- function(x) sum(getUniques(x))\n",
    "track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))\n",
    "colnames(track) <- c(\"input\", \"filtered\", \"denoisedF\", \"denoisedR\", \"merged\", \"nonchim\")\n",
    "rownames(track) <- sample.names\n",
    "head(track)\n",
    "sum(track[,\"input\"])\n",
    "sum(track[,\"nonchim\"])\n",
    "```\n",
    "\n",
    "```{r Assign Taxonomy}\n",
    "taxa_GTDB <- assignTaxonomy(seqtab.nochim, \"GTDB_ssu_all_r214_cleaned_ready.fa.gz\", multithread = TRUE)\n",
    "taxa.print <- taxa_GTDB\n",
    "rownames(taxa.print) <- NULL\n",
    "head(taxa_GTDB)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fcc4807-6656-4feb-9534-f331f059782b",
   "metadata": {},
   "source": [
    "### Construct Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92d73314-4adc-440a-b3f5-6e501bcf4738",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences <- getSequences(seqtab.nochim)\n",
    "names(sequences) <- sequences\n",
    "alignment <- AlignSeqs(DNAStringSet(sequences), anchor = NA)\n",
    "phang.align <- phyDat(as(alignment, \"matrix\"), type = \"DNA\")\n",
    "dm <- dist.ml(phang.align)\n",
    "treeNJ <- NJ(dm)\n",
    "fit <- pml(treeNJ, data = phang.align)\n",
    "fitGTR <- update(fit, k = 4, inv = 0.2)\n",
    "\n",
    "save.image(file = \"20231021-AldabraGiantTortoises-dada2.RData\")\n",
    "\n",
    "fitGTR <- optim.pml(fitGTR,\n",
    "                    model = \"GTR\", optInv = TRUE, optGamma = TRUE,\n",
    "                    rearrangement = \"stochastic\", control = pml.control(trace = 0)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "411a4b4a-b87f-4311-b9bb-92dc6013cf45",
   "metadata": {},
   "source": [
    "### Add Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96445ca2-50f6-4bf2-b02f-436a539bd78f",
   "metadata": {},
   "outputs": [],
   "source": [
    "samdf <- read.table(\"20230113 sample_metadata.tsv\", sep = \"\\t\", header = T, row.names = 1, comment = \"\")\n",
    "samdf$AgeGroup <- gsub(\" (\", \"\\n(\", samdf$AgeGroup, fixed = TRUE)\n",
    "head(samdf)\n",
    "# Preparing phyloseq\n",
    "ps <- phyloseq(\n",
    "  otu_table(seqtab.nochim, taxa_are_rows = FALSE),\n",
    "  sample_data(samdf),\n",
    "  tax_table(taxa_GTDB), phy_tree(fitGTR$tree)\n",
    ")\n",
    "# Storing the DNA sequences of our ASVs in the refseq slot of the phyloseq object,then renaming our taxa to a short string. That way, the short new taxa names will appear in tables and plots, and we can still recover the DNA sequences corresponding to each ASV as needed with refseq(ps)\n",
    "dna <- Biostrings::DNAStringSet(taxa_names(ps))\n",
    "names(dna) <- taxa_names(ps)\n",
    "ps <- merge_phyloseq(ps, dna)\n",
    "taxa_names(ps) <- paste0(\"ASV\", seq(ntaxa(ps)))\n",
    "\n",
    "sample_data(ps)$Sex <- factor(sample_data(ps)$Sex, levels = c(\"Not determined\", \"F\", \"M\"))\n",
    "sample_data(ps)$EnvironmentGroup5 <- factor(sample_data(ps)$EnvironmentGroup5, levels = EnvironmentGroup5_sorted)\n",
    "\n",
    "# Adding the ASV ids from ASV table into the taxa table at the end\n",
    "taxa_names(ps) <- paste0(\"ASV_\", rownames(tax_table(ps)))\n",
    "taxic <- as.data.frame(ps@tax_table) # this will help in setting large color options\n",
    "taxic$ASV <- rownames(taxic)\n",
    "colnames(taxic)\n",
    "taxmat <- as.matrix(taxic)\n",
    "new.tax <- tax_table(taxmat)\n",
    "tax_table(ps) <- new.tax\n",
    "tax_table(ps)[tax_table(ps)[, \"Family\"] == \"\", \"Family\"] <- \"Unclassified family\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f75ea7cc-6efb-443e-8332-902a13bcd7d6",
   "metadata": {},
   "source": [
    "### Subset Phyloseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "156b0a9d-51ca-41fc-b661-235ecb25ac1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_bg_cr <- subset_samples(ps, EnvironmentGroup5 == \"Botanical\\nGarden\" | EnvironmentGroup5 == \"Curieuse\")\n",
    "ps_wild <- subset_samples(ps, EnvironmentGroup5 == \"Curieuse\" )\n",
    "ps_controlled <- subset_samples(ps, EnvironmentGroup5 != \"Curieuse\")\n",
    "ps_bg <- subset_samples(ps, EnvironmentGroup5 == \"Botanical\\nGarden\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6177b2-b49a-445f-85e4-14bbb6bc417d",
   "metadata": {},
   "source": [
    "### Phyloseq Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83696dca-13a4-4f67-acaf-21c765998e9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "summarize_phyloseq(ps)\n",
    "print(ps)\n",
    "length(unique(tax_table(ps)[,'Phylum']))\n",
    "length(unique(tax_table(ps)[,'Class']))\n",
    "length(unique(tax_table(ps)[,'Family']))\n",
    "length(unique(tax_table(ps)[,'Genus']))\n",
    "length(unique(tax_table(ps)[,'ASV']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96cfb0bf-7a0b-4b62-8094-662455bf9de2",
   "metadata": {},
   "source": [
    "# Save R Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eb3f3d0-65a4-4e95-a3ce-18862e8b4282",
   "metadata": {},
   "outputs": [],
   "source": [
    "save.image(file = R_image)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4637cce6-1761-4229-8d65-3b95170ca595",
   "metadata": {},
   "source": [
    "# Alpha Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74e25e28-3781-49d2-b4a7-a64c34e72b74",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Alpha diversity - whole dataset\n",
    "ps_df_rich <- alpha_diversity_calculation(ps)\n",
    "ps_df_rich$EnvironmentGroup5 <- factor(ps_df_rich$EnvironmentGroup5, levels = EnvironmentGroup5_sorted)\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich, \"EnvironmentGroup5\", \"Shannon\", \"Curieuse\", \"AgeGroup13\", mycols_agegroup13, \"\\nEnvironment\", \"Shannon\", \"Alpha Diversity of Aldabra Giant Tortoises from Different Envrionemnts\", \"Age Group\")\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich, \"EnvironmentGroup5\", \"InvSimpson\", \"Curieuse\", \"AgeGroup13\", mycols_agegroup13, \"\\nEnvironment\", \"Inverted Simpson\", \"Alpha Diversity of Aldabra Giant Tortoises from Different Envrionemnts\", \"Age Group\")\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich, \"EnvironmentGroup5\", \"Shannon\", \"Curieuse\", \"AgeGroup13\", mycols_agegroup13, \"\\n環境\", \"Shannon\", \"様々な環境のアルダブラゾウガメの\\n腸内細菌のアルファ多様性（Shannon)\", \"年齢層\")\n",
    "\n",
    "\n",
    "# Alpha diversity - wild\n",
    "ps_df_rich_wild <- alpha_diversity_calculation(ps_wild)\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_wild, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[3], \"\\nAge Group\", \"Shannon\", \"Alpha Diversity (Shannon) of Aldabra Giant Tortoises from Wild Envrionemnt\", \"Environment\")\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_wild, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[3], \"\\n年齢層\", \"Shannon\", \"様々な年齢層の野生アルダブラゾウガメの\\n腸内細菌のアルファ多様性（Shannon)\", \"環境\")\n",
    "\n",
    "\n",
    "# Alpha diversity - controlled\n",
    "ps_df_rich_controlled <- alpha_diversity_calculation(ps_controlled)\n",
    "ps_df_rich_controlled$EnvironmentGroup5 <- factor(ps_df_rich_controlled$EnvironmentGroup5, levels = EnvironmentGroup5_sorted[c(1,2,4,5)])\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_controlled, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[c(1,2,4,5)], \"\\nAge Group\", \"Shannon\", \"Alpha Diversity (Shannon) of Aldabra Giant Tortoises from Controlled Envrionemnts\", \"Environment\")\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_controlled, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[c(1,2,4,5)], \"\\n年齢層\", \"Shannon\", \"様々な年齢層の飼育アルダブラゾウガメの\\n腸内細菌のアルファ多様性（Shannon)\", \"環境\")\n",
    "\n",
    "\n",
    "# Alpha diversity - Botanical Garden\n",
    "ps_df_rich_bg <- alpha_diversity_calculation(ps_bg)\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_bg, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[4], \"\\nAge Group\", \"Shannon\", \"Alpha Diversity (Shannon) of Aldabra Giant Tortoises from Botanical Garden\", \"Environment\")\n",
    "\n",
    "\n",
    "# Alpha diversity - Botanical Gardn + Curieuse\n",
    "ps_df_rich_bg_cr <- alpha_diversity_calculation(ps_bg_cr)\n",
    "ps_df_rich_bg_cr$EnvironmentGroup5 <- factor(ps_df_rich_bg_cr$EnvironmentGroup5, levels = EnvironmentGroup5_sorted[c(3,4)])\n",
    "\n",
    "alpha_diversity_plot(ps_df_rich_bg_cr, \"AgeGroup13\", \"Shannon\", \"100\",  \"EnvironmentGroup5\", mycols_environment5[c(3,4)], \"\\nAge Group\", \"Shannon\", \"Alpha Diversity (Shannon) of Aldabra Giant Tortoises from Controlled Envrionemnts\", \"Environment\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28ee751d-cdf5-4119-a068-96de51890d64",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Beta Diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32058dbb-506d-4f0d-a1f6-554c16d62af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "beta_ordination_plot(ps_rel_abund, \"pcoa\", \"bray curtis\")\n",
    "\n",
    "beta_ordination_plot(ps_rel_abund, \"pcoa\", \"unweighted unifrac\")\n",
    "\n",
    "beta_ordination_plot(ps_rel_abund, \"pcoa\", \"weighted unifrac\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04cafe1d-5149-4fd2-af3b-846197f220e9",
   "metadata": {},
   "source": [
    "# Relative Abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9309ae4e-f015-4bfd-a225-842e177e45b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ps_rel_abund <- transform_sample_counts(ps, function(otu) {otu * 100 / sum(otu, na.rm = T)})\n",
    "ps_rel_abund_wild <- transform_sample_counts(ps_wild, function(otu) {otu * 100 / sum(otu, na.rm = T)})\n",
    "ps_rel_abund_controlled <- transform_sample_counts(ps_controlled, function(otu) {otu * 100 / sum(otu, na.rm = T)}) \n",
    "\n",
    "# Phylum\n",
    "ps_rel_abund_phylumcore_df <- ps_rel_abund_taxacore_createdf(ps_rel_abund, \"phylum\")\n",
    "ps_rel_abund_taxacore_plot(ps_rel_abund_phylumcore_df, \"phylum\")\n",
    "ps_rel_abund_taxacore_0.01remainder_createdf_plot(ps_rel_abund_phylumcore_df,\"phylum\")\n",
    "\n",
    "# Family\n",
    "ps_rel_abund_familycore_df <- ps_rel_abund_taxacore_createdf(ps_rel_abund, \"family\")\n",
    "ps_rel_abund_taxacore_0.01remainder_createdf_plot(ps_rel_abund_familycore_df,\"family\")\n",
    "\n",
    "# Genera\n",
    "ps_rel_abund_genuscore_df <- ps_rel_abund_taxacore_createdf(ps_rel_abund, \"genus\")\n",
    "ps_rel_abund_taxacore_0.01remainder_createdf_plot(ps_rel_abund_genuscore_df,\"genus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c32ced13-f282-414c-bcea-b0580964ede1",
   "metadata": {},
   "source": [
    "# Venn Diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eecc54e1-04e1-4566-850f-3547f47825f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Venn diagram for the 4 main age groups\n",
    "venn_taxa_plot(ps_rel_abund, \"AgeGroup\", AgeGroup_vennsorted, \"ASVs\")\n",
    "venn_taxa_plot(ps_rel_abund, \"AgeGroup\", AgeGroup_vennsorted, \"Genus\")\n",
    "venn_taxa_plot(ps_rel_abund, \"AgeGroup\", AgeGroup_vennsorted, \"Species\")\n",
    "venn_taxa_plot(ps_rel_abund, \"AgeGroup\", AgeGroup_vennsorted, \"Phylum\")\n",
    "venn_taxa_plot(ps_rel_abund, \"AgeGroup\", AgeGroup_vennsorted, \"Family\")\n",
    "\n",
    "venn_taxa_plot(ps_rel_abund_wild, \"AgeGroup\", AgeGroup_vennsorted_wild, \"ASVs\")\n",
    "venn_taxa_plot(ps_rel_abund_wild, \"AgeGroup\", AgeGroup_vennsorted_wild, \"Genus\")\n",
    "venn_taxa_plot(ps_rel_abund_wild, \"AgeGroup\", AgeGroup_vennsorted_wild, \"Species\")\n",
    "\n",
    "venn_taxa_plot(ps_rel_abund_controlled, \"AgeGroup\", AgeGroup_vennsorted_controlled, \"Genus\")\n",
    "venn_taxa_plot(ps_rel_abund_controlled, \"AgeGroup\", AgeGroup_vennsorted_controlled, \"Species\")\n",
    "venn_taxa_plot(ps_rel_abund_controlled, \"AgeGroup\", AgeGroup_vennsorted_controlled, \"ASVs\")\n",
    "\n",
    "# Venn diagram for the 4 main environment groups\n",
    "venn_taxa_plot(ps_rel_abund, \"EnvironmentGroup5\",EnvironmentGroup5_vennsorted, \"ASVs\")\n",
    "venn_taxa_plot(ps_rel_abund, \"EnvironmentGroup5\",EnvironmentGroup5_vennsorted, \"Genus\")\n",
    "venn_taxa_plot(ps_rel_abund, \"EnvironmentGroup5\",EnvironmentGroup5_vennsorted,\"Species\")\n",
    "venn_taxa_plot(ps_rel_abund, \"EnvironmentGroup5\",EnvironmentGroup5_vennsorted,\"Phylum\")\n",
    "venn_taxa_plot(ps_rel_abund, \"EnvironmentGroup5\",EnvironmentGroup5_vennsorted,\"Family\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
